feat: Selectable non-editable text in blocks without inline content

matthewlipski opened the pull request:
This PR makes it possible to have blocks without inline content which contain selectable, non-editable text. Currently, you cannot select text while the block is selected, and selecting text while a different block is selected in the editor leads to a mismatch in editor & DOM selections. You can toggle this functionality using the `canSelectText` entry in a block config.

matthewlipski committed:
Made it possible to select text in blocks without inline content

YousefED commented on packages/core/src/extensions/FullySelectedNode/FullySelectedNodeExtension.ts:
```diff
@@ -0,0 +1,72 @@
+  },
+  onDestroy() {
+    document.removeEventListener("selectionchange", () => {
+      onSelectionChange(this.editor);
```
this doesn't work (you're removing a newly created function)

YousefED commented on packages/core/src/extensions/FullySelectedNode/FullySelectedNodeExtension.ts:
```diff
@@ -0,0 +1,72 @@
+  );
+
+  // Selected block has no inline content.
+  const selectedNodeHasNoContent =
```
a) remove double negation
b) should we check whether the node has the option enabled?

YousefED commented on packages/core/src/extensions/FullySelectedNode/FullySelectedNodeExtension.ts:
```diff
@@ -0,0 +1,72 @@
+
+  // Node selection is active.
+  const isNodeSelection = "node" in editor.state.selection;
+  if (!isNodeSelection) {
```
this check seems easier than the code above. move to top? (even before document.getSelection?)

YousefED commented on packages/core/src/schema/blocks/types.ts:
```diff
@@ -60,6 +61,7 @@ export type BlockConfig =
       type: string;
       readonly propSchema: PropSchema;
       content: "inline" | "none" | "table";
+      canSelectText?: boolean;
```
maybe `allowTextSelection`?

YousefED commented on packages/core/src/editor/tiptap.css:
```diff
@@ -0,0 +1,77 @@
+}
+
+/* Edited section to replace `.ProseMirror-hideselection` with `.ProseMirror-fullyselected */
+.ProseMirror-fullyselected *::selection {
```
let's review this file together and maybe rename to `fullyselectednode` or sth

YousefED commented on packages/core/src/editor/Block.css:
```diff
@@ -268,11 +268,13 @@ NESTED BLOCKS
   flex-direction: column;
   justify-content: stretch;
-  user-select: none;
+}
```
why this change?

YousefED commented on packages/core/src/api/exporters/copyExtension.ts:
```diff
@@ -65,20 +27,35 @@ const copyToClipboard = <
   // for the use-case in which only a block without content is
   // selected, e.g. an image block.
-  if (
+  const fragment =
```
discuss file on call

YousefED commented on packages/core/src/schema/blocks/createSpec.ts:
```diff
@@ -61,6 +65,75 @@ export type CustomBlockImplementation<
+  nodeView: NodeView
+) {
+  // Necessary for DOM to handle selections.
+  nodeView.ignoreMutation = () => true;
```
shouldn't this only be the case for selection events?

YousefED commented on packages/core/src/schema/blocks/createSpec.ts:
```diff
@@ -61,6 +65,75 @@ export type CustomBlockImplementation<
+      return true;
+    }
+
+    // Keyboard events should be handled by the browser. This doesn't prevent
```
sure? what about backspace or delete?

YousefED commented on packages/core/src/schema/blocks/createSpec.ts:
```diff
@@ -61,6 +65,75 @@ export type CustomBlockImplementation<
+      // If the target element contains only text, the browser should handle
+      // the event to update the selection. Otherwise, the event should be
+      // handled by BlockNote to select the block.
+      return (
```
let's discuss

YousefED committed:
fix bug?

matthewlipski committed:
Implemented PR feedback & cleaned up code

matthewlipski committed:
- Refactored classes and CSS
- Fixed copying when whole node is selected
- Added formatting toolbar controller fix

YousefED committed:
Merge remote-tracking branch 'origin/main' into node-text-selection
